File: internal/inspect/detectors/resource_finder.go
package detectors

// Copyright (C) 2025 by Posit Software, PBC.

import (
	"bufio"
	"fmt"
	"io"
	"net/url"
	"os"
	"regexp"
	"slices"
	"strings"

	"github.com/posit-dev/publisher/internal/config"
	"github.com/posit-dev/publisher/internal/logging"
	"github.com/posit-dev/publisher/internal/util"
)

type ResourceFinder interface {
	FindResources() ([]ExternalResource, error)
}

type multiResourceFinderFactory func(log logging.Logger, base util.AbsolutePath, filesFromConfig []string) (ResourceFinder, error)

type ExternalResource struct {
	Path string // Relative path to the resource
}

type resourceFinder struct {
	log          logging.Logger
	base         util.AbsolutePath
	InputDir     util.AbsolutePath
	InputFile    util.AbsolutePath
	InputFileExt string
	Resources    map[string]ExternalResource
}

type multiResourceFinder struct {
	finders []ResourceFinder
}

func NewMultiResourceFinder(log logging.Logger, base util.AbsolutePath, filesFromConfig []string) (*multiResourceFinder, error) {
	// config.Files are relative to base, although they have absolute path notation in Config object (e.g: "/logo.png")
	var finders []ResourceFinder
	for _, file := range filesFromConfig {
		absPath := base.Join(file)
		rf, err := NewResourceFinder(log, base, absPath)
		if err != nil {
			// Files incompatible with resource finder, we just skip them
			log.Debug("Cannot build resource finder for file", "file", file, "error", err.Error())
			continue
		}
		finders = append(finders, rf)
	}
	return &multiResourceFinder{finders: finders}, nil
}

func (rf *multiResourceFinder) FindResources() ([]ExternalResource, error) {
	resourceMap := make(map[string]ExternalResource)
	for _, finder := range rf.finders {
		resources, err := finder.FindResources()
		if err != nil {
			return nil, err
		}
		for _, res := range resources {
			resourceMap[res.Path] = res
		}
	}

	var resources []ExternalResource
	for _, resource := range resourceMap {
		resources = append(resources, resource)
	}

	return resources, nil
}

func NewResourceFinder(log logging.Logger, base util.AbsolutePath, inputFile util.AbsolutePath) (*resourceFinder, error) {
	// Validate file type
	ext := strings.ToLower(inputFile.Ext())
	supportedExtensions := map[string]bool{
		".md":   true,
		".rmd":  true,
		".qmd":  true,
		".html": true,
		".htm":  true,
		".r":    true,
		".css":  true,
	}

	if !supportedExtensions[ext] {
		return nil, fmt.Errorf("resource discovery is only supported for R Markdown, Quarto files or HTML (and CSS) files")
	}

	return &resourceFinder{
		log:          log,
		base:         base,
		InputDir:     inputFile.Dir(),
		InputFile:    inputFile,
		InputFileExt: ext,
		Resources:    make(map[string]ExternalResource),
	}, nil
}

func (rf *resourceFinder) FindResources() ([]ExternalResource, error) {
	// Create an empty slice to collect resources
	var resources []ExternalResource

	rf.log.Debug("Starting resource discovery", "file", rf.InputFile.String())

	// Discover resources based on file type
	var err error
	switch rf.InputFileExt {
	case ".md", ".rmd", ".qmd":
		err = rf.discoverMarkdownResources(rf.InputFile)
	case ".html", ".htm":
		err = rf.discoverHTMLResources(rf.InputFile)
	case ".r":
		err = rf.discoverRResources(rf.InputFile)
	case ".css":
		err = rf.discoverCSSResources(rf.InputFile)
	}

	if err != nil {
		return nil, err
	}

	// Convert map to slice for return
	for _, resource := range rf.Resources {
		resources = append(resources, resource)
	}

	return resources, nil
}

// Add a resource to the finder if it exists on disk
func (rf *resourceFinder) addResource(relpath string, explicit bool) bool {
	var abspath util.AbsolutePath
	rf.log.Debug("Found resource, validating...", "resource", relpath)

	// Resources that start with / are relative to the project's base directory
	hasLeadingSlash := strings.HasPrefix(relpath, "/")
	if hasLeadingSlash {
		abspath = rf.base.Join(relpath)
	} else {
		abspath = rf.InputDir.Join(relpath)
	}

	// Check if path exists on disk
	if _, err := abspath.Stat(); os.IsNotExist(err) {
		rf.log.Debug("Resource does not exist on disk, ignoring", "resource", relpath)
		return false
	}

	// We only add directories if this is an explicit reference
	// (e.g. from resource_files: in YAML front matter)
	// Implicit references (e.g. from Markdown image links) must be files
	// so we skip directories in that case
	if !explicit {
		fi, err := abspath.Stat()
		if err != nil {
			return false
		}
		if fi.IsDir() {
			return false
		}
	}

	relToBase, err := abspath.Rel(rf.base)
	if err != nil {
		// Ignoring error here, issues making a relative path should not stop discovery
		return false
	}

	// If not already tracked, or if this is an explicit reference to something
	// that was previously discovered implicitly, add/update it
	relPathString := relToBase.String()
	if _, exists := rf.Resources[relPathString]; !exists {
		rf.log.Debug("Including resource", "resource", relPathString)
		rf.Resources[relPathString] = ExternalResource{
			// Be consistent with list items, always return relative paths without leading slash
			Path: strings.TrimPrefix(relPathString, "/"),
		}
	}

	// If this is an R, CSS, or HTML file, process it recursively
	ext := strings.ToLower(abspath.Ext())
	if ext == ".r" || ext == ".css" || ext == ".html" || ext == ".htm" {
		switch ext {
		case ".r":
			rf.discoverRResources(abspath)
		case ".css":
			rf.discoverCSSResources(abspath)
		case ".html", ".htm":
			rf.discoverHTMLResources(abspath)
		}
	}

	return true
}

func (rf *resourceFinder) processResourceMatches(matches [][]string, branchedFrom *util.AbsolutePath) {
	for _, match := range matches {
		if len(match) > 1 {
			path := match[1]
			if rf.isFullURL(path) {
				// Skip web URLs, we only want to add local resources
				continue
			}
			if branchedFrom != nil {
				// When a resource was found in a file that was found to be a resource itself,
				// we need to resolve the path relative to the inspection input dir
				fullPath := branchedFrom.Dir().Join(path)
				relPath, err := fullPath.Rel(rf.InputDir)
				if err != nil {
					// Ignoring error here, issues making a relative path should not stop discovery
					continue
				}
				path = relPath.String()
			}
			rf.addResource(path, false)
		}
	}
}

func (rf *resourceFinder) processResourceByteMatches(matches [][][]byte, branchedFrom *util.AbsolutePath) {
	stringMatches := make([][]string, len(matches))
	for i, match := range matches {
		stringMatch := make([]string, len(match))
		for j, m := range match {
			stringMatch[j] = string(m)
		}
		stringMatches[i] = stringMatch
	}
	rf.processResourceMatches(stringMatches, branchedFrom)
}

// Process markdown link matches and add only HTML files as resources
func (rf *resourceFinder) processHTMLLinkMatches(matches [][]string) {
	for _, match := range matches {
		if len(match) > 1 {
			path := match[1]
			if rf.isFullURL(path) {
				// Skip web URLs, we only want to add local resources
				continue
			}
			// Only include HTML files
			if strings.HasSuffix(strings.ToLower(path), ".html") || strings.HasSuffix(strings.ToLower(path), ".htm") {
				rf.addResource(path, false)
			}
		}
	}
}

// Discover resources in Markdown, R Markdown, and Quarto files
func (rf *resourceFinder) discoverMarkdownResources(fpath util.AbsolutePath) error {
	file, err := fpath.Open()
	if err != nil {
		rf.log.Debug("Could not open file for resource discovery", "file", fpath.String(), "error", err.Error())
		return err
	}
	defer file.Close()

	// Create a scanner to read the file line by line
	scanner := bufio.NewScanner(file)

	// Regular expression for Markdown image syntax: ![alt](path)
	imgRegex := regexp.MustCompile(`!\[.*?\]\((.*?)(?:\s+["'].*?["'])?\)`)

	// Regular expression for Markdown link syntax: [text](path)
	// This will also match images, but we filter for .html/.htm in processing
	linkRegex := regexp.MustCompile(`\[.*?\]\((.*?)(?:\s+["'].*?["'])?\)`)

	// Regular expression for inline HTML image tags: <img src="path" ...>
	htmlImgRegex := regexp.MustCompile(`<img\s+[^>]*src=["'](.*?)["'][^>]*>`)

	// Scan for YAML front matter and explicit resource declarations
	inYAML := false
	var yamlContent strings.Builder

	for scanner.Scan() {
		line := scanner.Text()

		// Check for YAML front matter delimiters
		if line == "---" {
			if !inYAML {
				inYAML = true
				continue
			} else {
				inYAML = false
				// Parse YAML content for resources
				rf.parseYAMLResourceFiles(yamlContent.String())
				yamlContent.Reset()
				continue
			}
		}

		if inYAML {
			yamlContent.WriteString(line)
			yamlContent.WriteString("\n")
			continue
		}

		// Find Markdown image references
		matches := imgRegex.FindAllStringSubmatch(line, -1)
		rf.processResourceMatches(matches, nil)

		// Find Markdown links to HTML files
		linkMatches := linkRegex.FindAllStringSubmatch(line, -1)
		rf.processHTMLLinkMatches(linkMatches)

		// Find HTML image tags
		htmlMatches := htmlImgRegex.FindAllStringSubmatch(line, -1)
		rf.processResourceMatches(htmlMatches, nil)
	}

	if err := scanner.Err(); err != nil {
		return err
	}

	return nil
}

// Discover resources in R files
func (rf *resourceFinder) discoverRResources(fpath util.AbsolutePath) error {
	file, err := fpath.Open()
	if err != nil {
		rf.log.Debug("Could not open file for resource discovery", "file", fpath.String(), "error", err.Error())
		return err
	}
	defer file.Close()

	content, err := io.ReadAll(file)
	if err != nil {
		rf.log.Debug("Could not read file for resource discovery", "file", fpath.String(), "error", err.Error())
		return err
	}

	// Remove comments, so we don't pick up possible paths in comments
	contentStr := string(content)
	lines := strings.Split(contentStr, "\n")
	for i, line := range lines {
		if idx := strings.Index(line, "#"); idx >= 0 {
			lines[i] = line[:idx]
		}
	}
	contentStr = strings.Join(lines, "\n")

	// Extract double-quoted strings
	doubleQuoteRegex := regexp.MustCompile(`"([^"\n]*)"`)
	matches := doubleQuoteRegex.FindAllStringSubmatch(contentStr, -1)
	rf.processResourceMatches(matches, &fpath)

	// Extract single-quoted strings
	singleQuoteRegex := regexp.MustCompile(`'([^'\n]*)'`)
	matches = singleQuoteRegex.FindAllStringSubmatch(contentStr, -1)
	rf.processResourceMatches(matches, &fpath)

	return nil
}

// Discover resources in HTML files
func (rf *resourceFinder) discoverHTMLResources(fpath util.AbsolutePath) error {
	file, err := fpath.Open()
	if err != nil {
		rf.log.Debug("Could not open file for resource discovery", "file", fpath.String(), "error", err.Error())
		return err
	}
	defer file.Close()

	content, err := io.ReadAll(file)
	if err != nil {
		rf.log.Debug("Could not read file for resource discovery", "file", fpath.String(), "error", err.Error())
		return err
	}

	// Regular expressions to find common resource references in HTML
	patterns := map[string]*regexp.Regexp{
		"img":    regexp.MustCompile(`<img\s+[^>]*src=["'](.*?)["'][^>]*>`),
		"link":   regexp.MustCompile(`<link\s+[^>]*href=["'](.*?)["'][^>]*>`),
		"script": regexp.MustCompile(`<script\s+[^>]*src=["'](.*?)["'][^>]*>`),
		"video":  regexp.MustCompile(`<video\s+[^>]*src=["'](.*?)["'][^>]*>`),
		"audio":  regexp.MustCompile(`<audio\s+[^>]*src=["'](.*?)["'][^>]*>`),
		"source": regexp.MustCompile(`<source\s+[^>]*src=["'](.*?)["'][^>]*>`),
	}

	// Apply each pattern to the content
	for _, pattern := range patterns {
		matches := pattern.FindAllSubmatch(content, -1)
		rf.processResourceByteMatches(matches, &fpath)
	}

	return nil
}

// Discover resources in CSS files
func (rf *resourceFinder) discoverCSSResources(fpath util.AbsolutePath) error {
	file, err := fpath.Open()
	if err != nil {
		rf.log.Debug("Could not open file for resource discovery", "file", fpath.String(), "error", err.Error())
		return err
	}
	defer file.Close()

	content, err := io.ReadAll(file)
	if err != nil {
		rf.log.Debug("Could not read file for resource discovery", "file", fpath.String(), "error", err.Error())
		return err
	}

	// Find URL references in CSS
	urlRegex := regexp.MustCompile(`url\(['"]?([^'")]+)['"]?\)`)
	matches := urlRegex.FindAllSubmatch(content, -1)
	rf.processResourceByteMatches(matches, &fpath)

	return nil
}

// Parses YAML front matter for resource_files entries
func (rf *resourceFinder) parseYAMLResourceFiles(yamlContent string) {
	rf.log.Debug("Parsing YAML front matter looking up for resources")

	// Look for resource_files: entries in YAML
	lines := strings.Split(yamlContent, "\n")
	inResourceFiles := false

	for _, line := range lines {
		// Trim whitespace
		trimmedLine := strings.TrimSpace(line)

		// Check for resource_files: section
		if strings.HasPrefix(trimmedLine, "resource_files:") {
			rf.log.Debug("Found resource_files spec in YAML front matter")
			inResourceFiles = true
			continue
		}

		// If we're in the resource_files section and the line starts with a dash,
		// it's likely a list item declaring a resource
		if inResourceFiles && strings.HasPrefix(trimmedLine, "-") {
			// Extract the path from the line (removing the dash and trimming)
			path := strings.TrimSpace(strings.TrimPrefix(trimmedLine, "-"))

			// If it's wrapped in quotes, remove them
			if (strings.HasPrefix(path, "'") && strings.HasSuffix(path, "'")) ||
				(strings.HasPrefix(path, "\"") && strings.HasSuffix(path, "\"")) {
				path = path[1 : len(path)-1]
			}

			if path != "" && !rf.isFullURL(path) {
				// Handle wildcards and directories
				if strings.Contains(path, "*") {
					// For wildcards, use filepath.Glob to find matching files
					matches, err := rf.InputDir.Glob(path)
					if err == nil {
						for _, match := range matches {
							relpath, _ := match.Rel(rf.InputDir)
							rf.addResource(relpath.String(), true)
						}
					}
				} else {
					// Check if it's a directory
					fullPath := rf.InputDir.Join(path)
					fi, err := fullPath.Stat()
					if err == nil && fi.IsDir() {
						// For directories, walk the directory and add all files
						fullPath.Walk(func(filePath util.AbsolutePath, info os.FileInfo, err error) error {
							if err == nil && !info.IsDir() {
								relpath, _ := filePath.Rel(rf.InputDir)
								rf.addResource(relpath.String(), true)
							}
							return nil
						})
					} else {
						// Otherwise add as a regular resource
						rf.addResource(path, true)
					}
				}
			}
		} else if inResourceFiles && !strings.HasPrefix(trimmedLine, " ") && trimmedLine != "" {
			// If we encounter a line that doesn't start with space and isn't empty,
			// we've left the resource_files section
			inResourceFiles = false
		}
	}
}

func (rf *resourceFinder) isFullURL(path string) bool {
	u, err := url.ParseRequestURI(path)
	return err == nil && u.Scheme != "" && u.Host != ""
}

// Use resource finder to identify additional resources for the configuration.
// Additional static assets can be scattered alongside files.
func findAndIncludeAssets(
	log logging.Logger,
	rfFactory multiResourceFinderFactory,
	base util.AbsolutePath,
	cfg *config.Config,
) {
	rFinder, err := rfFactory(log, base, cfg.Files)
	if err != nil {
		log.Error(fmt.Sprintf("Error creating resource finder for %s project", cfg.Type), "error", err)
		return
	}
	resources, err := rFinder.FindResources()
	if err != nil {
		log.Error(fmt.Sprintf("Error finding resources for %s project", cfg.Type), "error", err)
		return
	}
	for _, rsrc := range resources {
		// Do not include assets that are nested in an already included directory.
		// e.g. if /index_files is included, do not include /index_files/custom.css
		rsrcRoot := strings.Split(rsrc.Path, "/")[0]
		rsrcStringToAdd := fmt.Sprint("/", rsrc.Path)
		rsrcDirIncluded := slices.Contains(cfg.Files, fmt.Sprint("/", rsrcRoot))
		if !rsrcDirIncluded && !slices.Contains(cfg.Files, rsrcStringToAdd) {
			cfg.Files = append(cfg.Files, rsrcStringToAdd)
		}
	}
}

File: extensions/vscode/src/inspect/helpers/resourceFinder.ts
// Copyright (C) 2026 by Posit Software, PBC.

import * as fs from "fs/promises";
import * as path from "path";
import picomatch from "picomatch";
import { logger } from "src/logging";

// Supported file extensions for resource scanning
const supportedExtensions = new Set([
  ".md",
  ".rmd",
  ".qmd",
  ".html",
  ".htm",
  ".r",
  ".css",
]);

// Extensions that trigger recursive scanning when discovered as resources
const recursiveExtensions = new Set([".html", ".htm", ".css", ".r"]);

// Markdown image: ![alt](path) with optional title
const markdownImgRe = /!\[.*?\]\((.*?)(?:\s+["'].*?["'])?\)/g;
// Markdown link: [text](path) — filtered to .html/.htm only
const markdownLinkRe = /\[.*?\]\((.*?)(?:\s+["'].*?["'])?\)/g;
// Inline HTML img in markdown: <img src="path" ...>
const htmlImgInMarkdownRe = /<img\s+[^>]*src=["'](.*?)["'][^>]*>/g;

// HTML resource patterns
const htmlPatterns = [
  /<img\s+[^>]*src=["'](.*?)["'][^>]*>/g,
  /<link\s+[^>]*href=["'](.*?)["'][^>]*>/g,
  /<script\s+[^>]*src=["'](.*?)["'][^>]*>/g,
  /<video\s+[^>]*src=["'](.*?)["'][^>]*>/g,
  /<audio\s+[^>]*src=["'](.*?)["'][^>]*>/g,
  /<source\s+[^>]*src=["'](.*?)["'][^>]*>/g,
];

// CSS url() references
const cssUrlRe = /url\(['"]?([^'")]+)['"]?\)/g;

// R quoted strings
const doubleQuoteRe = /"([^"\n]*)"/g;
const singleQuoteRe = /'([^'\n]*)'/g;

/**
 * Discover linked resources (images, CSS, scripts, etc.) referenced from
 * content files. Scans file contents via regex and returns additional
 * paths to include in the deployment bundle.
 *
 * @param baseDir - Absolute path to the project root
 * @param files - Current file list (leading-"/" relative paths)
 * @returns Additional "/"-prefixed relative paths to append
 */
export async function findLinkedResources(
  baseDir: string,
  files: string[],
): Promise<string[]> {
  // Collect all discovered resources keyed by relative path for dedup
  const resourceMap = new Set<string>();
  const visited = new Set<string>();

  for (const file of files) {
    const absPath = path.join(baseDir, file);
    const ext = path.extname(absPath).toLowerCase();

    if (!supportedExtensions.has(ext)) {
      continue;
    }

    await discoverResources(baseDir, absPath, visited, resourceMap);
  }

  // Filter out resources already in the input files list or nested under
  // already-included directories
  const result: string[] = [];
  for (const relPath of resourceMap) {
    const prefixed = relPath.startsWith("/") ? relPath : `/${relPath}`;

    // Skip if already in input files
    if (files.includes(prefixed)) {
      continue;
    }

    // Skip if nested under an already-included directory
    const rootSegment = prefixed.split("/")[1];
    if (rootSegment && files.includes(`/${rootSegment}`)) {
      continue;
    }

    result.push(prefixed);
  }

  return result;
}

async function discoverResources(
  baseDir: string,
  absFilePath: string,
  visited: Set<string>,
  resourceMap: Set<string>,
): Promise<void> {
  const ext = path.extname(absFilePath).toLowerCase();

  let content: string;
  try {
    content = await fs.readFile(absFilePath, "utf-8");
  } catch {
    logger.debug(`[resourceFinder] could not read file: ${absFilePath}`);
    return;
  }

  switch (ext) {
    case ".md":
    case ".rmd":
    case ".qmd":
      await scanMarkdown(baseDir, absFilePath, content, visited, resourceMap);
      break;
    case ".html":
    case ".htm":
      await scanHTML(baseDir, absFilePath, content, visited, resourceMap);
      break;
    case ".css":
      await scanCSS(baseDir, absFilePath, content, visited, resourceMap);
      break;
    case ".r":
      await scanR(baseDir, absFilePath, content, visited, resourceMap);
      break;
  }
}

async function scanMarkdown(
  baseDir: string,
  absFilePath: string,
  content: string,
  visited: Set<string>,
  resourceMap: Set<string>,
): Promise<void> {
  const inputDir = path.dirname(absFilePath);
  const lines = content.split("\n");

  let inYAML = false;
  let yamlContent = "";
  let yamlParsed = false;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i] ?? "";
    if (line.trim() === "---") {
      if (!inYAML && !yamlParsed && i === 0) {
        // Opening --- must be the first line of the file
        inYAML = true;
        continue;
      } else if (inYAML) {
        inYAML = false;
        yamlParsed = true;
        await parseYAMLResourceFiles(
          yamlContent,
          baseDir,
          inputDir,
          visited,
          resourceMap,
        );
        yamlContent = "";
        continue;
      }
    }

    if (inYAML) {
      yamlContent += line + "\n";
      continue;
    }

    // Markdown image references
    const imgMatches = extractMatches(markdownImgRe, line);
    await processMatches(
      imgMatches,
      baseDir,
      inputDir,
      false,
      visited,
      resourceMap,
    );

    // Markdown links — only .html/.htm targets
    const linkMatches = extractMatches(markdownLinkRe, line);
    await processHTMLLinkMatches(
      linkMatches,
      baseDir,
      inputDir,
      visited,
      resourceMap,
    );

    // Inline HTML <img> tags
    const htmlImgMatches = extractMatches(htmlImgInMarkdownRe, line);
    await processMatches(
      htmlImgMatches,
      baseDir,
      inputDir,
      false,
      visited,
      resourceMap,
    );
  }
}

async function scanHTML(
  baseDir: string,
  absFilePath: string,
  content: string,
  visited: Set<string>,
  resourceMap: Set<string>,
): Promise<void> {
  const inputDir = path.dirname(absFilePath);

  for (const pattern of htmlPatterns) {
    const matches = extractMatches(pattern, content);
    await processMatches(
      matches,
      baseDir,
      inputDir,
      false,
      visited,
      resourceMap,
    );
  }
}

async function scanCSS(
  baseDir: string,
  absFilePath: string,
  content: string,
  visited: Set<string>,
  resourceMap: Set<string>,
): Promise<void> {
  const inputDir = path.dirname(absFilePath);
  const matches = extractMatches(cssUrlRe, content);
  await processMatches(matches, baseDir, inputDir, false, visited, resourceMap);
}

async function scanR(
  baseDir: string,
  absFilePath: string,
  content: string,
  visited: Set<string>,
  resourceMap: Set<string>,
): Promise<void> {
  const inputDir = path.dirname(absFilePath);

  // Strip comments before extracting strings
  const strippedLines = content.split("\n").map((line) => {
    const commentIdx = line.indexOf("#");
    return commentIdx >= 0 ? line.substring(0, commentIdx) : line;
  });
  const stripped = strippedLines.join("\n");

  const doubleMatches = extractMatches(doubleQuoteRe, stripped);
  await processMatches(
    doubleMatches,
    baseDir,
    inputDir,
    false,
    visited,
    resourceMap,
  );

  const singleMatches = extractMatches(singleQuoteRe, stripped);
  await processMatches(
    singleMatches,
    baseDir,
    inputDir,
    false,
    visited,
    resourceMap,
  );
}

// ---- Match processing ----

function extractMatches(regex: RegExp, text: string): string[] {
  const results: string[] = [];
  // Reset lastIndex since we reuse global regexes
  regex.lastIndex = 0;
  let match: RegExpExecArray | null;
  while ((match = regex.exec(text)) !== null) {
    if (match[1]) {
      results.push(match[1]);
    }
  }
  return results;
}

async function processMatches(
  matches: string[],
  baseDir: string,
  inputDir: string,
  explicit: boolean,
  visited: Set<string>,
  resourceMap: Set<string>,
): Promise<void> {
  for (const candidate of matches) {
    if (isWebURL(candidate)) {
      continue;
    }

    await addResource(
      baseDir,
      inputDir,
      candidate,
      explicit,
      visited,
      resourceMap,
    );
  }
}

async function processHTMLLinkMatches(
  matches: string[],
  baseDir: string,
  inputDir: string,
  visited: Set<string>,
  resourceMap: Set<string>,
): Promise<void> {
  for (const candidate of matches) {
    if (isWebURL(candidate)) {
      continue;
    }
    const lower = candidate.toLowerCase();
    if (lower.endsWith(".html") || lower.endsWith(".htm")) {
      await addResource(
        baseDir,
        inputDir,
        candidate,
        false,
        visited,
        resourceMap,
      );
    }
  }
}

// ---- Resource validation and addition ----

async function addResource(
  baseDir: string,
  inputDir: string,
  relpath: string,
  explicit: boolean,
  visited: Set<string>,
  resourceMap: Set<string>,
): Promise<void> {
  logger.debug(`[resourceFinder] found resource, validating: ${relpath}`);

  // Resolve absolute path: leading "/" means relative to baseDir
  let absPath: string;
  if (relpath.startsWith("/")) {
    absPath = path.join(baseDir, relpath);
  } else {
    absPath = path.resolve(inputDir, relpath);
  }

  // Check existence on disk
  let stat: { isFile(): boolean; isDirectory(): boolean };
  try {
    stat = await fs.stat(absPath);
  } catch {
    logger.debug(
      `[resourceFinder] resource does not exist on disk: ${relpath}`,
    );
    return;
  }

  // Only allow directories for explicit resource_files declarations
  if (!explicit && stat.isDirectory()) {
    return;
  }

  // Compute path relative to baseDir
  const relToBase = path.relative(baseDir, absPath).replace(/\\/g, "/");

  // Add to map if not already tracked
  if (!resourceMap.has(relToBase)) {
    logger.debug(`[resourceFinder] including resource: ${relToBase}`);
    resourceMap.add(relToBase);
  }

  // Recursively scan HTML, CSS, and R files
  const ext = path.extname(absPath).toLowerCase();
  if (recursiveExtensions.has(ext) && !visited.has(absPath)) {
    visited.add(absPath);
    await discoverResources(baseDir, absPath, visited, resourceMap);
  }
}

// ---- YAML resource_files parsing ----

async function parseYAMLResourceFiles(
  yamlContent: string,
  baseDir: string,
  inputDir: string,
  visited: Set<string>,
  resourceMap: Set<string>,
): Promise<void> {
  const lines = yamlContent.split("\n");
  let inResourceFiles = false;

  for (const line of lines) {
    const trimmed = line.trim();

    if (trimmed.startsWith("resource_files:")) {
      logger.debug(
        "[resourceFinder] found resource_files spec in YAML front matter",
      );
      inResourceFiles = true;
      continue;
    }

    if (inResourceFiles && trimmed.startsWith("-")) {
      let itemPath = trimmed.substring(1).trim();

      // Remove quotes
      if (
        (itemPath.startsWith("'") && itemPath.endsWith("'")) ||
        (itemPath.startsWith('"') && itemPath.endsWith('"'))
      ) {
        itemPath = itemPath.substring(1, itemPath.length - 1);
      }

      if (!itemPath || isWebURL(itemPath)) {
        continue;
      }

      if (itemPath.includes("*")) {
        // Glob pattern: expand matches
        await expandGlob(itemPath, baseDir, inputDir, visited, resourceMap);
      } else {
        // Check if it's a directory
        const fullPath = path.resolve(inputDir, itemPath);
        let isDir = false;
        try {
          const stat = await fs.stat(fullPath);
          isDir = stat.isDirectory();
        } catch {
          // Not found — try as regular file
        }

        if (isDir) {
          await walkDirectory(
            fullPath,
            baseDir,
            inputDir,
            visited,
            resourceMap,
          );
        } else {
          await addResource(
            baseDir,
            inputDir,
            itemPath,
            true,
            visited,
            resourceMap,
          );
        }
      }
    } else if (inResourceFiles && !trimmed.startsWith(" ") && trimmed !== "") {
      // Left the resource_files section
      inResourceFiles = false;
    }
  }
}

// ---- Glob expansion ----

async function expandGlob(
  pattern: string,
  baseDir: string,
  inputDir: string,
  visited: Set<string>,
  resourceMap: Set<string>,
): Promise<void> {
  const isMatch = picomatch(pattern);

  // Walk inputDir to find matching files
  const entries = await walkDirectoryEntries(inputDir);
  for (const relEntry of entries) {
    if (isMatch(relEntry)) {
      await addResource(
        baseDir,
        inputDir,
        relEntry,
        true,
        visited,
        resourceMap,
      );
    }
  }
}

async function walkDirectoryEntries(dir: string): Promise<string[]> {
  const results: string[] = [];

  async function walk(currentDir: string, prefix: string): Promise<void> {
    let entries: Array<{
      name: string;
      isFile(): boolean;
      isDirectory(): boolean;
    }>;
    try {
      entries = await fs.readdir(currentDir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      const relPath = prefix ? `${prefix}/${entry.name}` : entry.name;
      if (entry.isFile()) {
        results.push(relPath);
      } else if (entry.isDirectory()) {
        await walk(path.join(currentDir, entry.name), relPath);
      }
    }
  }

  await walk(dir, "");
  return results;
}

async function walkDirectory(
  dirPath: string,
  baseDir: string,
  inputDir: string,
  visited: Set<string>,
  resourceMap: Set<string>,
): Promise<void> {
  let entries: Array<{
    name: string;
    isFile(): boolean;
    isDirectory(): boolean;
  }>;
  try {
    entries = await fs.readdir(dirPath, { withFileTypes: true });
  } catch {
    return;
  }

  for (const entry of entries) {
    const fullPath = path.join(dirPath, entry.name);
    if (entry.isFile()) {
      const relToInput = path.relative(inputDir, fullPath);
      await addResource(
        baseDir,
        inputDir,
        relToInput,
        true,
        visited,
        resourceMap,
      );
    } else if (entry.isDirectory()) {
      await walkDirectory(fullPath, baseDir, inputDir, visited, resourceMap);
    }
  }
}

function isWebURL(s: string): boolean {
  return (
    s.startsWith("http://") ||
    s.startsWith("https://") ||
    s.startsWith("ftp://") ||
    s.startsWith("data:")
  );
}
